src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,238 @@
+    {
+        internal class AmazonS3ClientFactory
+        {
+            public virtual IAmazonS3 Create(AWSCredentials awsCredentials, AmazonS3Config configuration) => new AmazonS3Client(awsCredentials, configuration);
+        }
+
+        internal AmazonS3ClientFactory ClientFactory = new();
+
+        public S3StorageEgressProvider(ILogger<S3StorageEgressProvider> logger) : base(logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<CancellationToken, Task<Stream>> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await using var stream = await action(token);
+
+                var client = await CreateClientAsync(options, token);
+                if (stream.CanSeek) // use the stream directly
+                {
+                    var request = new PutObjectRequest
+                    {
+                        BucketName = options.BucketName,
+                        ContentType = artifactSettings.ContentType,
+                        Key = artifactSettings.Name,
+                        InputStream = stream,
+                        AutoCloseStream = false,
+                    };
+                    await client.PutObjectAsync(request, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    using TransferUtility transferUtility = new(client);
+                    await transferUtility.UploadAsync(stream, options.BucketName, artifactSettings.Name, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (AmazonS3Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<Stream, CancellationToken, Task> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                var client = await CreateClientAsync(options, token);
+                var uploadId = await InitMultiPartUploadAsync(client, options.BucketName, artifactSettings, token);
+                await using var stream = new MultiPartUploadStream(client, options.BucketName, artifactSettings.Name, uploadId, options.CopyBufferSize!.Value);
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await action(stream, token);
```

Comment:
If the artifact capture action happens to fail, does this egress provider need to somehow tell S3 to drop all of the parts or cancel the uploadId?

Reply:
https://docs.aws.amazon.com/sdkfornet1/latest/apidocs/html/M_Amazon_S3_AmazonS3Client_AbortMultipartUpload.htm